File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.cpp
class Solution {
public:
    string reverseWords(string s) {
       stringstream ss(s);
       string word;
       vector<string>words;
       while(ss >> word){
          words.push_back(word);
       }
       reverse(words.begin(), words.end());
       string finalString = "";
       for(int i=0; i<words.size(); i++){
            finalString+=words[i];
            if(i!=words.size()-1){
                finalString+=" ";
            }
       }
       return finalString;
    }
};

File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.java
class Solution {
    private void reverse(int start,int end, StringBuffer s){
        while(start < end){
            char temp =s.charAt(start);
            s.setCharAt(start,s.charAt(end));
            s.setCharAt(end,temp);
            start++;
            end--;
        }
    }
    public String reverseWords(String s) {
        int i=0;
        StringBuffer temp = new StringBuffer();

        while(i < s.length()){
            while(i<s.length() && s.charAt(i) ==' ') i++; //skipping leading zeroes;

            while(i < s.length() && s.charAt(i)!=' '){
                temp.append(s.charAt(i));
                i++;
            }

            while(i<s.length() && s.charAt(i)==' ') i++; //skipping the continuing zeroes;

            if(i <s.length()){
                temp.append(' ');  // adding a single space after each word;
            }
        }

        StringBuffer result = new StringBuffer(temp.toString());

        reverse(0,result.length()-1,result);
        int start = 0;
        for(int end =0;end<=result.length(); end++){

            if(end == result.length() || result.charAt(end)==' '){
                reverse(start,end-1,result);
                start = end+1;
            }
        }


        return result.toString();
    }
}